java/flight/flight-sql/src/main/java/org/apache/arrow/flight/sql/FlightSqlClient.java

```diff
@@ -1048,14 +1049,34 @@ private Schema deserializeSchema(final ByteString bytes) {
     public FlightInfo execute(final CallOption... options) {
       checkOpen();
 
-      final FlightDescriptor descriptor = FlightDescriptor
+      FlightDescriptor descriptor = FlightDescriptor
           .command(Any.pack(CommandPreparedStatementQuery.newBuilder()
                   .setPreparedStatementHandle(preparedStatementResult.getPreparedStatementHandle())
                   .build())
               .toByteArray());
 
       if (parameterBindingRoot != null && parameterBindingRoot.getRowCount() > 0) {
-        putParameters(descriptor, options);
+        SyncPutListener putListener = putParameters(descriptor, options);
+
+        try {
+          final PutResult read = putListener.read();
+          if (read != null) {
+            try (final ArrowBuf metadata = read.getApplicationMetadata()) {
+              final FlightSql.DoPutPreparedStatementResult doPutPreparedStatementResult =
+                      FlightSql.DoPutPreparedStatementResult.parseFrom(metadata.nioBuffer());
+              descriptor = FlightDescriptor
+                      .command(Any.pack(CommandPreparedStatementQuery.newBuilder()
+                                      .setPreparedStatementHandle(
+                                              doPutPreparedStatementResult.getPreparedStatementHandle())
+                                      .build())
+                              .toByteArray());
```

Comment:
I would take a closer look to see if this conversion to byte array then back to byte string is necessary. I think it can be skipped.

Reply:
I'm assuming that doPutPreparedStatementResult.getPreparedStatementHandle() is the ByteString you are referring to?
But doesn't the handle need to be part of a CommandPreparedStatementQuery, so don't we need to build that query anyway. And the Any.pack and .toByteArray() are to convert that query into the correct format for .command()?
This follows the same conversion as preparedStatementResult above.
So I couldn't find a way to avoid this unless you can elaborate?